### PR TITLE
Several improvements to make this more reusable outside of founding usecase

### PR DESCRIPTION
**Pull request by john-sharratt**
- Added a `clear` method that is useful for reusing the octree from a clear slate
- Added a `upsert` method that can handle the situation when an element already exists at the target position
- Hidden away the internal structures from external consumers of the crate
- Exposed an `iter` method that gives access to the elements without the need to access internal structures
- Added `len` method that will return the number of elements in the tree without the need to access internal structures
- Added `is_empty` method that will return true if the tree is empty without the need to access internal structures
- Added an automatic formatting settings when using visual code (on save it will reformat)
- Removed the dependency on unstable rust, will now compile on stable rust
- Optimizations for when the octree has no garbage
- Implemented a faster garbage engine that scales better when lots of garbage exists
- Tweaked the garbage system and the benchmarks and improved the performance
- Iterators now support reverse direction
- Octree is now clonable
- Implemented `into_iter`
- Added more Position trait implementations for Arc, Box and Cow

### CODE REVIEW

**Review by exor2008 — COMMENTED** [review]
Great job!
I will fix [bevy_tree](https://github.com/exor2008/oktree/blob/f82f23a2fc9a5f9356a057350a6564c7b88b63e8/examples/bevy_tree.rs) example, add some tests for the new functionality and then release it as a `v0.3.0`

File: src/pool.rs [hunk 1]
```diff
@@ -11,26 +11,98 @@ use crate::{
...
+impl<T> From<T> for PoolItem<T> {
+    fn from(item: T) -> Self {
+        PoolItem {
+            item,
+            garbage: false,
+        }
+    }
+}
+
+impl<T> Clone for PoolItem<T>
```
Comment (exor2008) [inline src/pool.rs:30]: Is the explicit `Clone` implementation intentional?

Seems to work fine too:
```
#[repr(align(8))]
#[derive(Clone)]
pub(crate) struct PoolItem<T> {
```

File: src/pool.rs [hunk 2]
```diff
@@ -11,26 +11,98 @@ use crate::{
...
+            garbage: self.garbage,
+        }
+    }
+}
+
 /// [`Pool`] data structure.
 ///
 /// When element is removed no memory deallocation happens.
 /// Removed elements are only marked as deleted and their memory could be reused.  
 pub struct Pool<T> {
```
Comment (exor2008) [inline src/pool.rs:46]: Is the explicit `Clone` implementation intentional?

Seems to work fine too:

```
#[derive(Clone)]
pub struct Pool<T> {
```

File: src/tree.rs [hunk 3]
```diff
@@ -278,20 +329,49 @@ where
...
+        self.elements.iter()
+    }
+
+    /// Returns an iterator over the elements in the tree.
+    pub fn into_iter(self) -> impl IntoIterator<Item = T> {
+        self.elements.into_iter()
+    }
+}
+
+impl<U, T> Clone for Octree<U, T>
```
Comment (exor2008) [inline src/tree.rs:403]: Is the explicit `Clone ` implementation intentional?

Seems to work fine too:
```
#[derive(Clone)]
pub struct Octree<U, T>
```

File: src/bevy_integration.rs [hunk 4]
```diff
@@ -115,6 +115,61 @@ where
         }
     }
 
+    /// Intersect [`Octree`] with a custom .
```
Comment (exor2008) [inline src/bevy_integration.rs]: with a custom **intersection closure**?

**Comment by exor2008** [comment]
FYI: what I running to make sure everything is fine.
I should probably configure precommits with it 🤭

tests
```
cargo test --all-targets --all-features --release
```

clippy
```
cargo clippy --all-targets --all-features
```

examples
```
cargo run --all-features --example simple
cargo run --all-features --example bevy_tree
```

benchmark
```
cargo bench --all-features
```

docs
```
cargo doc --no-deps --open --all-features
```

**Review by exor2008 — COMMENTED** [review]
Could explicit `Clone` implementation be replaced with macros derivation?

**Comment by john-sharratt** [comment]
Cool, thanks for the tips on tests to run and `Clone` optimizations, all done in the last commit.
Didn't do the GIT hooks though

**Comment by john-sharratt** [comment]
I also added a test to make sure none clonable stuff still works

**Comment by john-sharratt** [comment]
This ones kind of cool
https://github.com/exor2008/oktree/pull/1/commits/5d60c5e6cbe9f562f49536872b52850b27b857d9

Removes most of the recursive callbacks however I don't notice that much difference in performance improvement (only marginal) - the biggest benefit of it is to reduce possibility stack overflows and memory usage

**Comment by john-sharratt** [comment]
- Fixed many of the doctests so they now compile and run
- Added a `intercept_for_each_with` to avoid the memory allocation of the return buffer

**Comment by john-sharratt** [comment]
```sh
   Doc-tests oktree

running 21 tests
test src\pool.rs - pool::Pool<Node<U>> (line 103) ... ignored
test src\pool.rs - pool::Pool<Node<U>> (line 118) ... ignored
test src\pool.rs - pool::Pool<NodeId> (line 168) ... ignored
test src\pool.rs - pool::Pool<NodeId> (line 186) ... ignored
test src\pool.rs - pool::Pool<T> (line 134) ... ignored
test src\pool.rs - pool::Pool<T> (line 152) ... ignored
test src\tree.rs - tree::Octree<U,T>::insert (line 93) ... ok
test src\lib.rs - NodeId (line 213) ... ok
test src\bevy_integration.rs - bevy_integration::Octree<U,T>::extend_intersect_with (line 190) ... ok
test src\lib.rs - ElementId (line 251) ... ok
test src\bevy_integration.rs - bevy_integration (line 8) ... ok
test src\tree.rs - tree::Octree<U,T>::find (line 292) ... ok
test src\tree.rs - tree::Octree<U,T>::upsert (line 210) ... ok
test src\bevy_integration.rs - bevy_integration::Octree<U,T>::intersect_with (line 163) ... ok
test src\lib.rs - (line 67) ... ok
test src\bevy_integration.rs - bevy_integration::Octree<U,T>::ray_cast (line 84) ... ok
test src\bevy_integration.rs - bevy_integration::Octree<U,T>::intersect (line 320) ... ok
test src\tree.rs - tree::Octree<U,T>::remove (line 235) ... ok
test src\bevy_integration.rs - bevy_integration::Octree<U,T>::intersect_with_for_each (line 254) ... ok
test src\bevy_integration.rs - bevy_integration (line 45) ... ok
test src\bevy_integration.rs - bevy_integration (line 28) ... ok

test result: ok. 15 passed; 0 failed; 6 ignored; 0 measured; 0 filtered out; finished in 5.50s
```

**Comment by exor2008** [comment]
> Cool, thanks for the tips on tests to run and `Clone` optimizations, all done in the last commit. Didn't do the GIT hooks though

Nice. I'll add CI pipeline for the PRs later.

**Comment by exor2008** [comment]
> This ones kind of cool [5d60c5e](https://github.com/exor2008/oktree/commit/5d60c5e6cbe9f562f49536872b52850b27b857d9)
> 
> Removes most of the recursive callbacks however I don't notice that much difference in performance improvement (only marginal) - the biggest benefit of it is to reduce possibility stack overflows and memory usage

Yeah, I tried to get rid of recursion in `Tree::insert` method but didn't get performance improvement too.
After all, even if your tree has a size of 65536 the deepest recursion depth would be 16, which doesn't seems so bad.

**Comment by john-sharratt** [comment]
Yeah, recursion on the insert method looks challenging to implement without a refactor, its probably not worth the effort. The others though don't drop the performance at least and will give advantages in stack overflow problems on things like WASM which bevy compiles to for browser based builds.

**Comment by john-sharratt** [comment]
P.s. This is what I'm using your creation for if you want to reuse the integration bits...

(EDIT: Got rid of the lookup to make it faster)

```rust
use crate::GameWorldReality;
use oktree::{prelude::TUVec3, tree::Octree, Position};
use smallvec::SmallVec;

use super::*;

pub struct WorldVisibilityNode {
    pub pos: TUVec3<u64>,
    pub entities: SmallVec<[Entity; 2]>,
}
impl Position for WorldVisibilityNode {
    type U = u64;
    fn position(&self) -> TUVec3<u64> {
        self.pos
    }
}

#[derive(Component, Clone)]
pub struct WorldVisibility {
    pub octree: Octree<u64, WorldVisibilityNode>,
    pub entities: Vec<Entity>,
}
impl WorldVisibility {
    pub fn project(&mut self) -> (&mut Octree<u64, WorldVisibilityNode>, &mut Vec<Entity>) {
        (&mut self.octree, &mut self.entities)
    }
}

#[derive(Component, Default)]
pub struct WorldInheritedVisibility {}

pub fn update_world_visibility(
    world_reality: Res<GameWorldReality>,
    mut world_visibility: Query<(&mut Visibility, &mut WorldVisibility), With<WorldVisibility>>,
    mut child_visibilities: Query<
        &mut Visibility,
        (With<WorldInheritedVisibility>, Without<WorldVisibility>),
    >,
) {
    let frustum_min = world_reality.camera_canvas.world_frustum.min().position();
    let frustum_max = world_reality.camera_canvas.world_frustum.max().position();

    world_visibility
        .par_iter_mut()
        .for_each(|(mut vis, mut world_vis)| {
            let (octree, entities) = world_vis.project();

            entities.clear();
            octree.intersect_with_for_each(
                |aabb| {
                    aabb.max.x.min(frustum_max.x) > aabb.min.x.max(frustum_min.x)
                        && aabb.max.y.min(frustum_max.y) > aabb.min.y.max(frustum_min.y)
                        && aabb.max.z.min(frustum_max.z) > aabb.min.z.max(frustum_min.z)
                },
                |node| {
                    entities.extend(node.entities.iter());
                },
            );

            *vis = if entities.is_empty() {
                Visibility::Hidden
            } else {
                Visibility::Inherited
            };
        });

    child_visibilities.par_iter_mut().for_each(|mut vis| {
        *vis = Visibility::Hidden;
    });
    for (_, world_vis) in world_visibility.iter() {
        world_vis
            .entities
            .iter()
            .filter_map(|e| child_visibilities.get_mut(*e).ok())
            .for_each(|mut vis| {
                *vis = Visibility::Inherited;
            });
    }
}

```

**Comment by exor2008** [comment]
> P.s. This is what I'm using your creation for if you want to reuse the integration bits...
> 
> ```rust
> use std::{borrow::Cow, collections::HashMap};
> 
> use crate::GameWorldReality;
> use oktree::{tree::Octree, Position};
> use sim::model::ObjectOrParticle;
> use smallvec::SmallVec;
> 
> use super::*;
> 
> #[derive(Component, Clone)]
> pub struct WorldVisibility {
>     pub octree: Cow<'static, Octree<u64, ObjectOrParticle<'static>>>,
>     pub lookup: HashMap<oktree::ElementId, SmallVec<[Entity; 2]>>,
>     pub entities: Vec<Entity>,
> }
> impl WorldVisibility {
>     pub fn project(
>         &mut self,
>     ) -> (
>         &mut Cow<'static, Octree<u64, ObjectOrParticle<'static>>>,
>         &mut HashMap<oktree::ElementId, SmallVec<[Entity; 2]>>,
>         &mut Vec<Entity>,
>     ) {
>         (&mut self.octree, &mut self.lookup, &mut self.entities)
>     }
> }
> 
> #[derive(Component, Default)]
> pub struct WorldInheritedVisibility {}
> 
> pub fn update_world_visibility(
>     world_reality: Res<GameWorldReality>,
>     mut world_visibility: Query<(&mut Visibility, &mut WorldVisibility), With<WorldVisibility>>,
>     mut child_visibilities: Query<
>         &mut Visibility,
>         (With<WorldInheritedVisibility>, Without<WorldVisibility>),
>     >,
> ) {
>     let frustum_min = world_reality.camera_canvas.world_frustum.min().position();
>     let frustum_max = world_reality.camera_canvas.world_frustum.max().position();
> 
>     world_visibility
>         .par_iter_mut()
>         .for_each(|(mut vis, mut world_vis)| {
>             let (octree, lookup, entities) = world_vis.project();
> 
>             entities.clear();
>             octree.intersect_with_for_each(
>                 |aabb| {
>                     aabb.max.x.min(frustum_max.x) > aabb.min.x.max(frustum_min.x)
>                         && aabb.max.y.min(frustum_max.y) > aabb.min.y.max(frustum_min.y)
>                         && aabb.max.z.min(frustum_max.z) > aabb.min.z.max(frustum_min.z)
>                 },
>                 |element| {
>                     if let Some(element_entities) = lookup.get(&element) {
>                         element_entities.iter().for_each(|entity| {
>                             entities.push(*entity);
>                         });
>                     }
>                 },
>             );
> 
>             *vis = if entities.is_empty() {
>                 Visibility::Hidden
>             } else {
>                 Visibility::Inherited
>             };
>         });
> 
>     child_visibilities.par_iter_mut().for_each(|mut vis| {
>         *vis = Visibility::Hidden;
>     });
>     for (_, world_vis) in world_visibility.iter() {
>         for entity in world_vis.entities.iter() {
>             if let Ok(mut vis) = child_visibilities.get_mut(*entity) {
>                 *vis = Visibility::Inherited;
>             }
>         }
>     }
> }
> ```

Ah, so it some kind of voxel engine?
Well, that's awesome!
Drop me a project link?

Btw there is another great crate that could be useful for [building optimized meshes](https://crates.io/crates/block-mesh).
Don't know if it will be useful in your case but it looks promising.

File: src/tree.rs [hunk 5]
```diff
@@ -194,18 +202,44 @@ where
...
-    /// let c1 = DummyCell::new(TUVec3::new(1u8, 1, 1));
+    /// ```rust
+    /// use oktree::prelude::*;
+    ///
+    /// let mut tree = Octree::from_aabb(Aabb::new(TUVec3::splat(16), 16).unwrap());
+    /// let c1 = TUVec3u8::new(1u8, 1, 1);
     /// let c1_id = tree.insert(c1).unwrap();
     ///
-    /// assert_eq!(tree.remove(c1_id).is_ok())
+    /// assert_eq!(tree.remove(c1_id).is_ok(), true);
```
Comment (exor2008) [inline src/tree.rs]:
```suggestion
    /// assert!(tree.remove(c1_id).is_ok());
```

File: src/tree.rs [hunk 6]
```diff
@@ -247,51 +306,114 @@ where
...
+impl<U: Unsigned, T: Position<U = U>> std::iter::IntoIterator for Octree<U, T> {
+    type Item = T;
+    type IntoIter = PoolIntoIterator<T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.elements.into_iter()
+    }
+}
+
+impl<U, T> Clone for Octree<U, T>
```
Comment (exor2008) [inline src/tree.rs:403]: Gonna replace that with derive macro too?
```suggestion
#[derive(Clone)]
```

**Review by exor2008 — COMMENTED** [review]
Minor fixes request

**Comment by john-sharratt** [comment]
> 
> Ah, so it some kind of voxel engine? Well, that's awesome! Drop me a project link?
> 
> Btw there is another great crate that could be useful for [building optimized meshes](https://crates.io/crates/block-mesh). Don't know if it will be useful in your case but it looks promising.

https://x.com/nguyen_anti/status/1851402467667234918

**Comment by john-sharratt** [comment]
The world is huge and I'm using it for the fixed models, not quite voxel but not far off.

Screenshot 2024-11-21 120848

File: src/tree.rs (hunk 6 above)
Comment (john-sharratt) [reply src/tree.rs:403]: Yeah, this is what I feared, I'll need to undo macros and go back to explicit

```shell
   Compiling battle-cities v0.1.0 (C:\Users\johna\prog\battle-cities\crates\client)
error[E0277]: the trait bound `dynamic_game_world::update_world_visibility::WorldVisibilityNode: Clone` is not satisfied
  --> crates\client\src\controller\dynamic_game_world\update_world_visibility.rs:20:5
   |
18 | #[derive(Component, Clone)]
   |                     ----- in this derive macro expansion
19 | pub struct WorldVisibility {
20 |     pub octree: Octree<u64, WorldVisibilityNode>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `dynamic_game_world::update_world_visibility::WorldVisibilityNode`, which is required by `Octree<u64, dynamic_game_world::update_world_visibility::WorldVisibilityNode>: Clone`
   |
   = note: required for `Octree<u64, dynamic_game_world::update_world_visibility::WorldVisibilityNode>` to implement `Clone`
   = note: this error originates in the derive macro `Clone` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `dynamic_game_world::update_world_visibility::WorldVisibilityNode` with `#[derive(Clone)]`
   |
7  + #[derive(Clone)]
8  | pub struct WorldVisibilityNode {
   |

For more information about this error, try `rustc --explain E0277`.
error: could not compile `battle-cities` (bin "battle-cities") due to 1 previous error
PS C:\Users\johna\prog\battle-cities>
```

Comment (john-sharratt) [reply src/tree.rs:403]: Last commit fixes the Clone requirement, not mandatory anymore

**Comment by exor2008** [comment]
> > Ah, so it some kind of voxel engine? Well, that's awesome! Drop me a project link?
> > Btw there is another great crate that could be useful for [building optimized meshes](https://crates.io/crates/block-mesh). Don't know if it will be useful in your case but it looks promising.
> 
> https://x.com/nguyen_anti/status/1851402467667234918

Awesome.
Resembles Kenshi at first glance.

**Comment by exor2008** [comment]
Btw let me know when you think you are done, so I can perform an overall code review, ok?

**Comment by john-sharratt** [comment]
I think its done enough, i can start another PR for more things if needed
